x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -51,83 +55,178 @@
 
-            IndexRequest indexRequest = new IndexRequest(InferenceIndexConstants.LATEST_INDEX_NAME)
-                .opType(DocWriteRequest.OpType.CREATE)
-                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
-                .id(trainedModelConfig.getModelId())
-                .source(source);
-            executeAsyncWithOrigin(client, ML_ORIGIN, IndexAction.INSTANCE, indexRequest,
-                ActionListener.wrap(
-                    r -> listener.onResponse(true),
-                    e -> {
-                        logger.error(new ParameterizedMessage(
-                            "[{}] failed to store trained model for inference", trainedModelConfig.getModelId()), e);
-                        if (ExceptionsHelper.unwrapCause(e) instanceof VersionConflictEngineException) {
-                            listener.onFailure(new ResourceAlreadyExistsException(
-                                Messages.getMessage(Messages.INFERENCE_TRAINED_MODEL_EXISTS, trainedModelConfig.getModelId())));
-                        } else {
-                            listener.onFailure(
-                                new ElasticsearchStatusException(Messages.INFERENCE_FAILED_TO_STORE_MODEL,
-                                    RestStatus.INTERNAL_SERVER_ERROR,
-                                    e,
-                                    trainedModelConfig.getModelId()));
-                        }
-                    }));
-        } catch (IOException e) {
-            // not expected to happen but for the sake of completeness
-            listener.onFailure(new ElasticsearchParseException(
-                Messages.getMessage(Messages.INFERENCE_FAILED_TO_SERIALIZE_MODEL, trainedModelConfig.getModelId()),
-                e));
-        }
+        ActionListener<IndexResponse> putDefinitionListener = ActionListener.wrap(
+            r -> listener.onResponse(true),
+            e -> {
+                logger.error(new ParameterizedMessage(
+                    "[{}] failed to store trained model definition for inference", trainedModelConfig.getModelId()), e);
+                if (ExceptionsHelper.unwrapCause(e) instanceof VersionConflictEngineException) {
+                    listener.onFailure(new ResourceAlreadyExistsException(
+                        Messages.getMessage(Messages.INFERENCE_TRAINED_MODEL_EXISTS, trainedModelConfig.getModelId())));
+                } else {
+                    listener.onFailure(
+                        new ElasticsearchStatusException(Messages.INFERENCE_FAILED_TO_STORE_MODEL,
+                            RestStatus.INTERNAL_SERVER_ERROR,
+                            e,
+                            trainedModelConfig.getModelId()));
+                }
+            }
+        );
+
+        ActionListener<IndexResponse> putConfigListener = ActionListener.wrap(
+            r -> {
+                if (trainedModelConfig.getDefinition() != null) {
```

Comment:
Should we allow for definition to ever be null when we're storing a model? I can't think of a reason to do so. Even if we add support for updates, we should have a different method for it. Unless there's a reason I'm missing, I think we could check the definition is not null and throw if it is early on this method.

Reply:
I suppose we can make it more flexible in the future if ever necessary (or provide specially handling for internal use cases).